src/TrimNoOps.cpp

```diff
@@ -0,0 +1,403 @@
+        // condition is false or the if body is a no-op.
+        total_condition = make_and(total_condition, make_or(!op->condition, condition));
+        condition = const_true();
+        if (op->else_case.defined()) {
+            op->else_case.accept(this);
+            total_condition = make_and(total_condition, make_or(op->condition, condition));
+        }
+        condition = total_condition;
+    }
+
+    void visit(const Call *op) {
+        // Certain intrinsics that may appear in loops have side-effects. Most notably: image_store.
+        if (op->call_type == Call::Intrinsic &&
+            (op->name == Call::rewrite_buffer ||
+             op->name == Call::image_store ||
+             op->name == Call::copy_memory)) {
+            condition = const_false();
+        } else {
+            IRVisitor::visit(op);
+        }
+    }
+
+    template<typename LetOrLetStmt>
+    void visit_let(const LetOrLetStmt *op) {
+        IRVisitor::visit(op);
+        if (expr_uses_var(condition, op->name)) {
+            condition = Let::make(op->name, op->value, condition);
+        }
+    }
+
+    void visit(const LetStmt *op) {
+        visit_let(op);
+    }
+
+    void visit(const Let *op) {
+        visit_let(op);
+    }
+
+public:
+    Expr condition = const_true();
+};
+
+class SimplifyUsingBounds : public IRMutator {
+    struct ContainingLoop {
+        string var;
+        Interval i;
+    };
+    vector<ContainingLoop> containing_loops;
+
+    using IRMutator::visit;
+
+    // Can we prove a condition over the non-rectangular domain of the for loops we're in?
+    bool provably_true_over_domain(Expr test) {
+        debug(3) << "Attempting to prove: " << test << "\n";
+        for (size_t i = containing_loops.size(); i > 0; i--) {
+            // Because the domain is potentially non-rectangular, we
+            // need to take each variable one-by-one, simplifying in
+            // between to allow for cancellations of the bounds of
+            // inner loops with outer loop variables.
+            auto loop = containing_loops[i-1];
+            if (is_const(test)) {
+                break;
+            } else if (!expr_uses_var(test, loop.var)) {
+                continue;
+            }  else if (is_one(simplify(loop.i.min == loop.i.max)) && expr_uses_var(test, loop.var)) {
+                // If min == max then either the domain only has one correct value, which we
+                // can substitute directly.
+                test = common_subexpression_elimination(Let::make(loop.var, loop.i.min, test));
+            } else if (is_one(simplify(loop.i.min >= loop.i.max)) && expr_uses_var(test, loop.var)) {
+                // If min >= max then either the domain only has one correct value,
+                // or the domain is empty, which implies both min/max are true under
+                // the domain.
+                test = common_subexpression_elimination(Let::make(loop.var, loop.i.min, test) ||
```

Comment:
Not sure CSE is needed here either - test is duplicated, but it's immediately consumed by simplify below, which at worst will simplify test twice, which is probably cheaper than running CSE.